File: chat-spigot/src/main/java/gg/warcraft/chat/spigot/event/SpigotChatEventMapper.java
package gg.warcraft.chat.spigot.event;

import com.google.inject.Inject;
import gg.warcraft.chat.app.event.NativeAsyncPlayerChatEvent;
import gg.warcraft.monolith.api.core.EventService;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

import java.util.UUID;

public class SpigotChatEventMapper implements Listener {
    private final EventService eventService;

    @Inject
    public SpigotChatEventMapper(EventService eventService) {
        this.eventService = eventService;
    }

    @EventHandler
    public void onAsyncPlayerChatEvent(AsyncPlayerChatEvent event) {
        UUID playerId = event.getPlayer().getUniqueId();
        NativeAsyncPlayerChatEvent nativeEvent = new NativeAsyncPlayerChatEvent(playerId, event.getMessage());
        eventService.publish(nativeEvent);
        event.setCancelled(true);
    }
}

File: chat-spigot/src/main/scala/gg/warcraft/chat/spigot/SpigotChatEventMapper.scala
package gg.warcraft.chat.spigot

import gg.warcraft.chat.AsyncPlayerPreChatEvent
import gg.warcraft.monolith.api.core.event.EventService
import gg.warcraft.monolith.api.player.PlayerService
import org.bukkit.event.{EventHandler, Listener}

class SpigotChatEventMapper(implicit
    eventService: EventService,
    playerService: PlayerService
) extends Listener {
  @EventHandler
  def preChat(event: SpigotAsyncPlayerChatEvent): Unit = {
    val player = playerService.getPlayer(event.getPlayer.getUniqueId)
    val preEvent = AsyncPlayerPreChatEvent(player, event.getMessage)
    eventService.publish(preEvent)
    event.setCancelled(true)
  }
}
